app/services/strategy_marketplace_service.py

```
@@ -1479,106 +1479,156 @@ async def get_user_strategy_portfolio(self, user_id: str) -> Dict[str, Any]:
+        # Add method-level timeout for entire operation to protect chat responsiveness
         try:
-            async with asyncio.timeout(60.0):  # 60 second timeout for entire method
-                return await self._get_user_strategy_portfolio_impl(user_id)
+            async with asyncio.timeout(12.0):
+                portfolio = await self._get_user_strategy_portfolio_impl(user_id)
+                if portfolio.get("success"):
+                    return portfolio
+                # If we received a structured but unsuccessful payload, fall back to AI snapshot
+                return self._build_fallback_ai_strategy_portfolio(user_id, "portfolio_error")
         except asyncio.TimeoutError:
             self.logger.error("❌ Portfolio fetch timeout", user_id=user_id)
```

Ruff (0.13.1)

Use `logging.exception` instead of `logging.error`

Replace with `exception`

(TRY400)